huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go

```diff
@@ -0,0 +1,229 @@
+					validation.StringMatch(regexp.MustCompile("^[\u4e00-\u9fa5\\w-]*$"),
+						"Only letters, chinese characters, digits, underscores (_) and hyphens (-) are allowed."),
+					validation.StringLenBetween(1, 64),
+				),
+				Description: "The name of the private NAT gateway.",
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(0, 255),
+					validation.StringMatch(regexp.MustCompile(`^[^<>]*$`),
+						"The angle brackets (< and >) are not allowed."),
+				),
+				Description: "The description of the private NAT gateway.",
+			},
+			"spec": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "The specification of the private NAT gateway.",
+				ValidateFunc: validation.StringInSlice([]string{
+					PrivateSpecTypeSmall,
+					PrivateSpecTypeMedium,
+					PrivateSpecTypeLarge,
+					PrivateSpecTypeExtraLarge,
+				}, false),
+			},
+			"enterprise_project_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "The ID of the enterprise project to which the private NAT gateway belongs.",
+			},
+			"tags": common.TagsSchema(),
+			"created_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The creation time of the private NAT gateway.",
+			},
+			"updated_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The latest update time of the private NAT gateway.",
+			},
+			"status": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The current status of the private NAT gateway.",
+			},
+		},
+	}
+}
+
+func buildDownLinkVpcs(subnetId string) []gateways.DownLinkVpc {
+	return []gateways.DownLinkVpc{
+		{
+			SubnetId: subnetId,
+		},
+	}
+}
+
+func resourcePrivateGatewayCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	client, err := cfg.NatV3Client(cfg.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating NAT v3 client: %s", err)
+	}
+
+	opts := gateways.CreateOpts{
+		Name:                d.Get("name").(string),
+		DownLinkVpcs:        buildDownLinkVpcs(d.Get("subnet_id").(string)),
+		Description:         d.Get("description").(string),
+		Spec:                d.Get("spec").(string),
+		EnterpriseProjectId: cfg.GetEnterpriseProjectID(d),
+		Tags:                utils.ExpandResourceTags(d.Get("tags").(map[string]interface{})),
+	}
+
+	log.Printf("[DEBUG] The create options of the private NAT gateway is: %#v", opts)
+	resp, err := gateways.Create(client, opts)
+	if err != nil {
+		return diag.Errorf("error creating private NAT gateway: %s", err)
+	}
+	d.SetId(resp.ID)
+
+	return resourcePrivateGatewayRead(ctx, d, meta)
+}
+
+func flattenDownLinkVpcs(vpcs []gateways.DownLinkVpc) string {
+	if len(vpcs) < 1 {
+		return ""
+	}
+	return vpcs[0].SubnetId
```

Comment:
you can use utils.Search()

Reply:
No function names `Search` in the `utils` package.

Reply:
done. The function `PathSearch` is useful.